src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3134,18 +3109,177 @@ void MacroAssembler::check_klass_subtype_slow_path(Register Rsubklass,
+void MacroAssembler::check_klass_subtype_slow_path_table(Register sub_klass,
+                                                         Register super_klass,
+                                                         Register temp_reg,
+                                                         Register temp2_reg,
+                                                         Register temp3_reg,
+                                                         Register temp4_reg,
+                                                         Register result_reg,
+                                                         Label* L_success,
+                                                         Label* L_failure,
+                                                         bool set_cond_codes) {
+  BLOCK_COMMENT("check_klass_subtype_slow_path_table {");
+
+  RegSet temps = RegSet::of(temp_reg, temp2_reg, temp3_reg, temp4_reg);
+
+  assert_different_registers(sub_klass, super_klass, temp_reg, temp2_reg, temp4_reg);
+
+  Label L_fallthrough;
+  int label_nulls = 0;
+  if (L_success == nullptr)   { L_success   = &L_fallthrough; label_nulls++; }
+  if (L_failure == nullptr)   { L_failure   = &L_fallthrough; label_nulls++; }
+  assert(label_nulls <= 1, "at most one null in the batch");
+
+  RegSetIterator<Register> available_regs
+  // Z_R0 will be used to hold Z_R15(Z_SP) while pushing a new frame, So don't use that here.
+  // Z_R1 is will be used to hold r_bitmap in lookup_secondary_supers_table_var, so can't be used
+  // Z_R2, Z_R3, Z_R4 will be used in secondary_supers_verify, for the failure reporting
+    = (RegSet::range(Z_R0, Z_R15) - temps - sub_klass - super_klass - Z_R1_scratch - Z_R0_scratch - Z_R2 - Z_R3 - Z_R4).begin();
+
+  RegSet pushed_regs;
+
+  int slots = 0;
+  constexpr int NumRegs = 5;
+  Register* regs[NumRegs] = {&temp_reg, &temp2_reg, &temp3_reg, &temp4_reg, &result_reg};
+  bool is_noreg[NumRegs] = {
+    temp_reg == noreg,
+    temp2_reg == noreg,
+    temp3_reg == noreg,
+    temp4_reg == noreg,
+    result_reg == noreg
+  };
+
+  // Allocate registers and count slots
+  for (int i = 0; i < NumRegs; ++i) {
+    if (is_noreg[i]) {
+      slots++;
+      *regs[i] = allocate_if_noreg(*regs[i], available_regs, pushed_regs);
+    }
+  }
+
+#ifdef ASSERT
+  assert(temp_reg   != noreg, "temp_reg: sanity");
+  assert(temp2_reg  != noreg, "temp2_reg: sanity");
+  assert(temp3_reg  != noreg, "temp3_reg: sanity");
+  assert(temp4_reg  != noreg, "temp4_reg: sanity");
+  assert(result_reg != noreg, "result_reg: sanity");
+#endif // ASSERT
+
+  const int frame_size = slots * BytesPerWord + frame::z_abi_160_size;
+
+  // Push & save registers
+  {
+    int i = 0;
+    save_return_pc();
+    push_frame(frame_size);
+
+    for (int j = 0; j < NumRegs; ++j) {
+      if (is_noreg[j]) {
+        z_stg(*regs[j], (i++) * BytesPerWord + frame::z_abi_160_size, Z_SP);
+      }
+    }
+
+    assert(i * BytesPerWord + frame::z_abi_160_size == frame_size, "sanity");
+  }
```

Comment:
push & pop could be moved to new method, likewise done by other archs. I need a second opinion on that as it looks like current solution is also okish.